lib/segment/src/vector_storage/simple_vector_storage.rs

```diff
@@ -166,13 +188,19 @@ impl VectorStorage for SimpleVectorStorage {
         for point_id in other_ids {
             check_process_stopped(stopped)?;
             // Do not perform preprocessing - vectors should be already processed
-            let other_deleted = other.is_deleted_vector(point_id);
             let other_vector = other.get_vector(point_id);
+            let other_deleted = other.is_deleted_vector(point_id);
             let new_id = self.vectors.push(other_vector)?;
-            self.set_deleted(new_id, other_deleted);
+            if other_deleted {
+                self.grow_deleted_and_set(new_id, true);
+            }
             self.update_stored(new_id, other_deleted, Some(other_vector))?;
         }
         let end_index = self.vectors.len() as PointOffsetType;
+
+        // Have enough deleted flags for all vectors
+        self.grow_deleted();
```

Comment:
I propose to grow deleted vectors before `for` statement. It's possible because the amount of deleted flags is a sum of `self.deleted_vector_bitslice.len()` and `other.deleted_vector_bitslice.len()`. Such change will decrease the reallocations amount because calling of `grow_deleted_and_set` will not reallocate bitvec

Reply:
We do not always take all IDs from `other`, and IDs are ultimately selected by `other_ids`. Because of that we can't use `other.deleted_vector_bitslice`.

That means we'd have to count this iterator first, which is worse than the current approach in my opinion.

What do you think?

Reply:
On second thought, maybe we can use that approach, grow it much larger in some cases and just don't care. After the for loop we could truncate to the actual number of vectors if we'd like.

What do you think?

Reply:
Updated it to use this approach. Grow to maximum size before the for loop. Then truncate after to the exact number of vectors.

Reply:
I guess truncating is fine because it still reduces the allocations count In the general case. You were right about `other.deleted_vector_bitslice` - we cannot use it. Because I approved this PR, I don't have any concerns about this proposal. If you find these changes too complicated, I'm ok with commit reversion or keeping it as is

Reply:
I think I'll keep it as is. This definitely improves preventing a lot of reallocations if we have many deleted vectors.

> I guess truncating is fine because it still reduces the allocations count In the general case

Truncating does [**not**](https://docs.rs/bitvec/latest/bitvec/vec/struct.BitVec.html#method.truncate) deallocate. It just shrinks the vector length. I don't think this is a big deal though, cases with a high difference are very unlikely and this is relatively small anyway.

Reply:
Yeah, sorry for the confusion. Agreed about rare cases of high difference. What do you think about calling `shrink_to_fit` after truncation?

Reply:
The only objection I have against it is that for multiple `update_from` with `shrink_to_fit` rounds we would re-and-de-allocate if the number of vectors is smaller than we expect. While with `truncate` we would reuse already allocated memory.

I assume this to be an exceptionally rare case, so I don't have a strong preference for either of the two.

Reply:
Ok, no doubts